### PR TITLE
Add git remote rename subcommand

### PR DESCRIPTION
**Pull request by Ralith**
I regularly reach for this when, upon deciding to contribute to a project, I rename `origin` to `upstream` and add a new `origin` addressing my fork.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

File: src/commands.rs [hunk 1]
```diff
@@ -3967,6 +3977,20 @@ fn cmd_git_remote_remove(
...
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &GitRemoteRenameArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let repo = workspace_command.repo();
+    let git_repo = get_git_repo(repo.store())?;
+    git_repo
+        .remote_rename(&args.old, &args.new)
+        .map_err(|err| CommandError::UserError(err.to_string()))?;
```
Comment (Ralith) [inline src/commands.rs:3990]: `jj git remote remove` manually checks for the remote, but the operation already contains that check and produces a pleasant enough error: `Error: remote 'asdf' does not exist; class=Config (7); code=NotFound (-3)`. Are we intentionally replacing that error with our own, perhaps to exclude the weird trailing details?

**Comment by Ralith** [comment]
Are there any references in jj's state to a remote by name that need to be updated?

**Comment by martinvonz** [comment]
> Are there any references in jj's state to a remote by name that need to be updated?
> 
> e: Found one, maybe?

Yep, that looks correct (the find and the fix).

FYI, I've considered identifying remotes by a uuid and then having a separate a the user-visible name in repo-level config. That would mean that we don't need to change anything in the repo when we rename a remote, which helps when operating on the operation log (undo/restore, concurrent operations). Currently (after this PR), if you rename a remote and then undo, we're going to undo the rename in the repo object but not in the config (in the git repo). I'm not sure it's worth the trouble to change it. Also, the remote's name is shared with the underlying git repo (when using the git backend, which is the only backend that currently supports push/pull), so it would be confusing if we renamed the remote to a uuid there.

File: src/commands.rs [hunk 2]
```diff
@@ -3967,6 +3977,31 @@ fn cmd_git_remote_remove(
...
+    let any_targets_affected = repo
+        .view()
+        .branches()
+        .values()
+        .any(|target| target.remote_targets.contains_key(&args.old));
+    if any_targets_affected {
```
Comment (martinvonz) [inline src/commands.rs:3996]: This check can be replaced by checking `tx.mut_repo().has_changes()` before calling `workspace_command.finish_transaction()` below. That's a little simpler, and I'm not concerned about the cost of unnecessarily starting a new transaction in the error case.

**Review by martinvonz — COMMENTED** [review]
Thanks!

Could you add a test case? There's a `test_git_remotes.rs` where I think it would belong.

File: src/commands.rs (hunk 1 above)
Comment (Ralith) [reply src/commands.rs:3990]: Makes sense to me!

File: src/commands.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply src/commands.rs:3996]: I was curious, actually; thanks!

**Comment by martinvonz** [comment]
> > Currently (after this PR), if you rename a remote and then undo, we're going to undo the rename in the repo object but not in the config (in the git repo). I'm not sure it's worth the trouble to change it.
> 
> We need to either use stable names or implement undos properly, don't we? Having `undo` corrupt the synchronization with the git repo sounds really bad. Is this solved for other git remote operations?

No, not solved for other operations either. For example, `jj git remote remove origin` will remove the `origin` remote from the git config and the remote branches from the repo. If you then undo that operation, you'll have remote branches without an associated remote in the config. Those remote branches will just sit there until you add back the remote and `jj git remote remove` it again. That's obviously not great, but at least it won't make the repo unusable (AFAIK).

**Comment by Ralith** [comment]
Ah, then maybe it's best left for future work?

**Comment by martinvonz** [comment]
> Ah, then maybe it's best left for future work?

Yep, I think so.

**Comment by martinvonz** [comment]
Btw, this is still unresolved:

> Could you add a test case? There's a `test_git_remotes.rs` where I think it would belong.

I should probably have left that as a comment on some file so it would actually appear as unresolved in the GitHub UI (not just logically unresolved). I'm still getting used to it.

**Review by martinvonz — APPROVED** [review]
Thanks!

**Comment by Ralith** [comment]
Yeah, just ran out of time last night. Test added now, plus another for the new lib API.

File: lib/tests/test_mut_repo.rs [hunk 3]
```diff
@@ -612,3 +612,18 @@ fn test_rebase_descendants_conflicting_rewrite(use_git: bool) {
...
+    let test_repo = TestRepo::init(use_git);
+    let repo = &test_repo.repo;
+    let mut tx = repo.start_transaction("test");
+    let mut_repo = tx.mut_repo();
+    let commit = testutils::create_random_commit(&settings, repo).write_to_repo(mut_repo);
+    let target = RefTarget::Normal(commit.id().clone());
+    mut_repo.set_remote_branch("main".to_string(), "origin".to_string(), target.clone());
+    mut_repo.rename_remote("origin", "upstream");
+    assert_eq!(mut_repo.get_remote_branch("main", "upstream"), Some(target));
+}
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply lib/tests/test_mut_repo.rs:629]: Thanks for adding this! Might be good to also check that the branch doesn't also exist under the old name.

[the remark this answers is not in the document]
Comment (Ralith) [reply lib/tests/test_mut_repo.rs:630]: Done.

**Pull request merged**